File: combinatorics/resource/ST/project_5/bai_toan_7/Top-down.py
from collections import deque

# Lop dinh nghia cay
class TreeNode:
    def __init__(self, label):
        self.label = label                  # Luu nhan cua nut
        self.children = []                 # Danh sach cac nut con

# Ham duyet top-down bang queue (level-order)
def topdown_traversal(root):
    if root is None:
        return
    q = deque()
    q.append(root)
    while q:
        node = q.popleft()                # Lay nut dau hang doi
        print(node.label)                 # In nhan cua nut
        for child in node.children:
            q.append(child)               # Dua cac nut con vao hang doi

# Ham nhap cay tu ban phim
def input_tree():
    label = input("Nhap nhan nut: ")
    node = TreeNode(label)
    n = int(input(f"So con cua {label}: "))
    for i in range(n):
        print(f"-- Con thu {i+1} cua {label}:")
        child = input_tree()
        node.children.append(child)
    return node

# Ham chinh
if __name__ == "__main__":
    print("Nhap cay:")
    root = input_tree()
    print("Duyet top-down:")
    topdown_traversal(root)

File: combinatorics/resource/ST/project_5/bai_toan_7/Top-down.cpp
#include <iostream>
#include <vector>
#include <string>
#include <queue>
using namespace std;

// Dinh nghia mot nut cay
struct TreeNode {
    string label;                         // Nhan cua nut
    vector<TreeNode*> children;          // Danh sach cac nut con
    TreeNode(string l) : label(l) {}     // Ham tao
};

// Ham duyet top-down bang hang doi
void topdownTraversal(TreeNode* root) {
    if (!root) return;
    queue<TreeNode*> q;
    q.push(root);                        // Dua nut goc vao hang doi
    while (!q.empty()) {
        TreeNode* node = q.front(); q.pop();     // Lay nut dau hang doi
        cout << node->label << endl;             // In ra nhan
        for (TreeNode* child : node->children)
            q.push(child);                       // Dua con vao hang doi
    }
}

// Ham nhap cay tu ban phim
TreeNode* inputTree() {
    string label;
    cout << "Nhap nhan nut: ";
    cin >> label;
    TreeNode* node = new TreeNode(label);
    int n;
    cout << "So con cua " << label << ": ";
    cin >> n;
    for (int i = 0; i < n; ++i) {
        cout << "-- Con thu " << i+1 << " cua " << label << ":\n";
        node->children.push_back(inputTree());
    }
    return node;
}

// Ham main
int main() {
    cout << "Nhap cay:\n";
    TreeNode* root = inputTree();
    cout << "Duyet top-down:\n";
    topdownTraversal(root);
    return 0;
}